### PR TITLE
Fix limited tracing

### PR DESCRIPTION
**Pull request by SammyK**
### Description

This PR removes the need to detect limited-tracing mode using `dd_trace_tracer_is_limited()` calls in tracing closures. This allows the tracer to stop tracing much earlier in the execution when the tracer is in limited mode; this in turn provides a much more optimized experience in limited mode.

Currently, to ensure `foo()` will not be instrumented in limited mode, an integration would make a call to `dd_trace_tracer_is_limited()` at the top of the tracing closure: 

```php
dd_trace_function('foo', function (\DDTrace\SpanData $span) {
    if (dd_trace_tracer_is_limited()) {
        return false;
    }
    // ...
});
```

With this PR, when limited mode is active, no calls will be instrumented unless `instrument_when_limited` is set to `1` from a new declarative API. To ensure a function or method is still instrumented even in limited tracing mode, pass a config array in place of a tracing closure to `dd_trace_function()` or `dd_trace_method()`.

For functions:

```php
dd_trace_function('foo', [
    'instrument_when_limited' => 1,
    'posthook' => function (\DDTrace\SpanData $span) {
        // ...
    }
]);
```

For methods:

```php
dd_trace_method('MyClass', 'myMethod', [
    'instrument_when_limited' => 1,
    'posthook' => function (\DDTrace\SpanData $span) {
        // ...
    }
]);
```

The only two supported config array keys in this PR are `posthook` (the tracing closure) and `instrument_when_limited` (signifies a call should be instrumented even when the tracer is in limited mode). More options can be added in the future to support a more declarative approach to integrations.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the appropriate release draft. Create one if necessary.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.
- [ ] Changelog has been added to the appropriate release draft. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: src/ext/logging.c [hunk 1]
```diff
@@ -2,13 +2,19 @@
...
+
+void _ddtrace_log_errf(const char *format, ...) {
     va_list args;
     char *buffer;
 
     va_start(args, format);
     vspprintf(&buffer, 0, format, args);
-    ddtrace_log_err(buffer);
+    TSRMLS_FETCH();
+    php_log_err(buffer TSRMLS_CC);
```
Comment (morrisonlevi) [inline src/ext/logging.c]: Why not just call `ddtrace_log_err`?

File: src/ext/logging.c [hunk 2]
```diff
@@ -2,13 +2,19 @@
 
 #include <php.h>
 
-void _ddtrace_log_errf(TSRMLS_FC const char *format, ...) {
+void ddtrace_log_err(char *message) {
```
Comment (morrisonlevi) [inline src/ext/logging.c]: I would put this in the header as `inline` (not static inline) and then here in the .c file put the `extern inline` declaration. I've done this is some places already; search for `extern inline` for examples.

File: src/ext/span.c [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #include <unistd.h>
 
 #include "ddtrace.h"
+#include "random.h"
```
Comment (morrisonlevi) [inline src/ext/span.c]: Why include random.h?

Comment (SammyK) [reply src/ext/span.c]: This is for the `ddtrace_free_span_id_stack()` call.

File: src/ext/logging.c (hunk 1 above)
Comment (SammyK) [reply src/ext/logging.c]: Nice catch! 👍

File: src/ext/logging.c (hunk 2 above)
Comment (SammyK) [reply src/ext/logging.c]: Excellent feedback. Updated! :)

File: src/ext/logging.h [hunk 4]
```diff
@@ -4,19 +4,21 @@
 
 #include "configuration.h"
 
-#if defined(ZTS) && PHP_VERSION_ID < 70000
-#define TSRMLS_FC TSRMLS_D,
-#define ddtrace_log_errf(...) _ddtrace_log_errf(TSRMLS_C, __VA_ARGS__)
-#else
-#define TSRMLS_FC
 #define ddtrace_log_errf(...) _ddtrace_log_errf(__VA_ARGS__)
```
Comment (morrisonlevi) [inline src/ext/logging.h]: Why do we have this macro that just forwards to one of our own functions?

Comment (SammyK) [reply src/ext/logging.h]: It made sense when we had separate API's under the hood for each PHP version, but now that we've refactored it quite a bit, we can certainly get rid of it. :)

**Pull request merged**